File: src/test/java/com/uapp_llc/test/stub/repository/AbstractRepositoryStub.java
package com.uapp_llc.test.stub.repository;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Predicate;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

abstract class AbstractRepositoryStub<ID, T> {

  protected final Logger logger = LoggerFactory.getLogger(this.getClass());
  private final Map<ID, T> entities = new LinkedHashMap<>();

  public T save(ID id, T entity) {
    Objects.requireNonNull(id, "Id must not be null");
    Objects.requireNonNull(entity, "Entity must not be null");
    boolean isNew = null == entities.put(id, entity);
    logger.debug("Entity with id '{}' {}: {} ", id, (isNew ? "added" : "updated"), entity);
    return entity;
  }

  public T find(ID id) {
    Objects.requireNonNull(id, "Id must not be null");
    T entity = entities.get(id);
    logger.debug("Entity with id '{}' {}found", id, (entity == null ? "not " : ""));
    return entity;
  }

  public T find(Predicate<T> criteria) {
    Objects.requireNonNull(criteria, "Criteria must not be null");
    return entities.values()
        .stream()
        .filter(criteria)
        .findFirst()
        .orElse(null);
  }

  public boolean exists(ID id) {
    Objects.requireNonNull(id, "Id must not be null");
    boolean exists = entities.containsKey(id);
    logger.debug("Entity with id '{}' {}exists", id, (exists ? "" : "not "));
    return exists;
  }

  public boolean exists(Predicate<T> criteria) {
    Objects.requireNonNull(criteria, "Criteria must not be null");
    return entities.values()
        .stream()
        .anyMatch(criteria);
  }

  public void delete(ID id, T entity) {
    Objects.requireNonNull(id, "Id must not be null");
    Objects.requireNonNull(entity, "Entity must not be null");
    boolean removed = entities.remove(id, entity);
    logger.debug("Entity with id '{}' {}removed: {}", id, (removed ? "" : "not "), entity);
  }

  public void deleteAll() {
    int size = entities.size();
    entities.clear();
    logger.debug("Delete {} entities", size);
  }

  public int size() {
    int size = entities.size();
    logger.debug("Repository size: {}", size);
    return size;
  }

  public List<T> findAll() {
    Collection<T> all = entities.values();
    logger.debug("{} entities found", all.size());
    return new ArrayList<>(all);
  }

}

File: src/test/kotlin/com/uapp_llc/test/stub/repository/AbstractRepositoryStub.kt
package com.uapp_llc.test.stub.repository

import org.slf4j.Logger
import org.slf4j.LoggerFactory

abstract class AbstractRepositoryStub<ID, T> {

  protected val logger: Logger = LoggerFactory.getLogger(this::class.java)
  private val entities: MutableMap<ID?, T> = mutableMapOf()

  open fun save(id: ID?, entity: T): T {
    val isNew: Boolean = entities.containsKey(id)
    entities[id] = entity
    logger.debug("Entity with id '{}' {}: {} ", id, (if (isNew) "added" else "updated"), entity)
    return entity
  }

  open fun find(id: ID?): T? {
    val entity: T? = entities[id]
    logger.debug("Entity with id '{}' {}found", id, (if (entity == null) "not " else ""))
    return entity
  }

  open fun find(criteria: (T) -> Boolean): T? {
    return entities.values
        .find(criteria)
  }

  open fun exists(id: ID?): Boolean {
    val exists: Boolean = entities.containsKey(id)
    logger.debug("Entity with id '{}' {}exists", id, (if (exists) "" else "not "))
    return exists
  }

  open fun exists(criteria: (T) -> Boolean): Boolean {
    return entities.values
        .any(criteria)
  }

  open fun delete(id: ID?, entity: T) {
    val removed: Boolean = entities.remove(id, entity)
    logger.debug("Entity with id '{}' {}removed: {}", id, (if (removed) "" else "not "), entity)
  }

  open fun deleteAll() {
    val size: Int = entities.size
    entities.clear()
    logger.debug("Delete {} entities", size)
  }

  open fun size(): Int {
    val size: Int = entities.size
    logger.debug("Repository size: {}", size)
    return size
  }

  open fun findAll(): MutableList<T> {
    val all: MutableCollection<T> = entities.values
    logger.debug("{} entities found", all.size)
    return all.toMutableList()
  }

}
